src/Common/src/System/IO/PathInternal.Windows.cs

```diff
@@ -232,114 +229,92 @@ internal static bool HasIllegalCharacters(string path, bool checkAdditional = fa
         /// <summary>
         /// Only check for ? and *.
         /// </summary>
-        internal static bool HasAdditionalIllegalCharacters(string path)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal unsafe static bool HasAdditionalIllegalCharacters(string path)
         {
             int startIndex = PathInternal.IsExtended(path) ? ExtendedPathPrefix.Length : 0;
-            return path.IndexOfAny(AdditionalInvalidPathChars, startIndex) >= 0;
+
+            char currentChar;
+            for (int i = startIndex; i < path.Length; i++)
+            {
+                currentChar = path[i];
+                if (currentChar == '*' || currentChar == '?') return true;
+            }
+            return false;
```

Comment:
@ianhays made the opposite change at https://github.com/dotnet/corefx/pull/3516 in the name of improved performance.  Before we change this and go back and forth on what's most efficient for this case, can the two of you confer and understand what you're doing differently in your benchmarks?